src/settings/file_watch.ts

```diff
@@ -20,100 +19,110 @@ function isInFolder(file: TFile, dir: TFolder) {
+		|| fileIsInFolder(plugin, plugin.settings.snippetVariablesFileLocation, file) 
+		|| fileIsInFolder(plugin, plugin.settings.snippetsFileLocation, file)
+	) {
 		try {
-			await refreshSnippetsFromFileOrFolder(plugin);
+			refreshFromFileOrFolder(plugin);
 		}
 		catch {
-			new Notice("Failed to load snippets.", 5000);
+			new Notice("Failed to load snippet/variable files.", 5000);
 		}
 	}
 }
 
-export const onFileCreate = (plugin: LatexSuitePlugin, file:TAbstractFile) => {
-	if (!(plugin.settings.loadSnippetsFromFile)) return;
+export const onFileCreate = (plugin: LatexSuitePlugin, file: TAbstractFile) => {
+	if (!(file instanceof TFile)) return;
 
-	if (file instanceof TFile && fileIsInSnippetsFolder(plugin, file)) {
-		refreshSnippetsFromFileOrFolder(plugin);
+	if (plugin.settings.loadSnippetVariablesFromFile && fileIsInFolder(plugin,plugin.settings.snippetVariablesFileLocation, file) 
+		|| plugin.settings.loadSnippetsFromFile && fileIsInFolder(plugin,plugin.settings.snippetsFileLocation, file)
+	) {
+		refreshFromFileOrFolder(plugin);
 	}
 }
 
-export const onFileDelete = (plugin: LatexSuitePlugin, file:TAbstractFile) => {
-	if (!(plugin.settings.loadSnippetsFromFile)) return;
+export const onFileDelete = (plugin: LatexSuitePlugin, file: TAbstractFile) => {
+	if (!(file instanceof TFile)) return;
 
+	const snippetVariablesDir = plugin.app.vault.getAbstractFileByPath(plugin.settings.snippetVariablesFileLocation);
 	const snippetDir = plugin.app.vault.getAbstractFileByPath(plugin.settings.snippetsFileLocation);
-	const isFolder = snippetDir instanceof TFolder;
 
-	if (file instanceof TFile && (isFolder && file.path.contains(snippetDir.path))) {
-		refreshSnippetsFromFileOrFolder(plugin);
+	if (plugin.settings.loadSnippetVariablesFromFile && snippetVariablesDir instanceof TFolder && file.path.contains(snippetVariablesDir.path)
+	 	|| plugin.settings.loadSnippetsFromFile && snippetDir instanceof TFolder && file.path.contains(snippetDir.path)
+	) {
+		refreshFromFileOrFolder(plugin);
 	}
 }
 
-async function getSnippetsFromFile(vault: Vault, file: TFile, snippetVariables: SnippetVariables) {
+async function getFromFile<T>(vault: Vault, file: TFile, parse: (content: string) => Promise<T>) {
 	const content = await vault.cachedRead(file);
-	let snippets:Snippet[] = [];
 
 	try {
-		snippets = await parseSnippets(content, snippetVariables);
+		return await parse(content);
 	}
 	catch (e) {
-		new Notice(`Failed to load snippet file ${file.name}`);
-		console.log(`Failed to load snippet file ${file.path}:`, e);
+		new Notice(`Failed to load snippet/variable file ${file.name}`);
+		console.log(`Failed to load snippet/variable file ${file.path}:`, e);
 	}
-
-	return snippets;
 }
 
-async function getSnippetsWithinFolder(vault: Vault, folder: TFolder, snippetVariables: SnippetVariables) {
-	const snippets:Snippet[] = [];
-
+async function* getWithinFolder<T>(vault: Vault, folder: TFolder, parse: (content: string) => Promise<T>): AsyncGenerator<T> {
```

Comment:
Rather than return a value once, a generator `yield`s any number of values consecutively:
```javascript
function* foo() {
    yield 5;
    yield 1;
    yield 3;
}

for (const generated of foo()) {
    console.log(generated)
}

// 5
// 1
// 3
```

The `yield*` keyword makes a parent generator yield all values from a child generator:
```javascript
function* foo() {
    yield 5;
    yield 1;
    yield 3;
}

function* bar() {
    yield "start";
    yield* foo(); // read as `for (const generated of foo()) { yield generated; }`
    yield "end";
}

for (const generated of bar()) {
    console.log(generated)
}

// start
// 5
// 1
// 3
// end
```

In our case:

The old code was collecting the imported objects with `Object.assign` or `Array.concat` at each step of the recursion. This means each helper function must know which one to use.

Using generators, the values from all recursions are "streamed" into a top level loop which collects them. This means only the top level functions care if we're reading snippets or variables.

Reply:
I see. Thank you for the explanation!

I'm personally in favour of changing the existing code from a recursive approach to a DFS/BFS-like approach, where we take in a `fileOrFolder: TAbstractFile` as input, traverse that folder, and return all the `TFiles` within it as an array/set. However, I'll leave it up to you if you feel that generator functions are the better way to go about things!